### DISCUSSION TITLE
Cold shutdown causes task loss

### DISCUSSION THREAD

**Discussion by kbiedrzycki** [celery/celery#8583] [Q&A]
Hey y'all. 

Recently I started adding more telemetry and monitoring around my celery cluster and I've noticed that some of the tasks are being lost in the event of shutdown. My celery stack is deployed to AWS using ECS Fargate as a platform. Also, I'm using `REMAP_SIGTERM=SIGQUIT` setting to make sure my celery workers are handling signals properly (I want them to shutdown using `cold` method). My workers are started using following command:

```
celery worker --without-gossip --without-mingle -A appname -l info -E --max-tasks-per-child=15 --concurrency=1 -Ofair -Qqueue -nxyz.worker%h
```

celery report:
```
software -> celery:4.4.7 (cliffs) kombu:4.6.11 py:3.9.16
            billiard:3.6.4.0 redis:4.3.4
platform -> system:Darwin arch:64bit
            kernel version:22.6.0 imp:CPython
loader   -> celery.loaders.default.Loader
settings -> transport:rediss results:disabled
```

So, what exactly I've noticed. While going through the logs, I've spotted events where celery was receiving shutdown event `worker: Cold shutdown (MainProcess)` (for example due to scaling activity, release or spot instance being interrupted). Usually, when this happens, I can also see `Restoring 1 unacknowledged message(s)` log that happens right after. I run all of my tasks with `acks_late=True` and `reject_on_worker_lost=True`. However, even with that setting, when cold shutdown event is received, some tasks are being lost. I can definitely see events where celery picks up task (`Received task` log occurs from `celery.worker.strategy` logger). Once it is picked up, after few seconds there's a shutdown event. I would expect task to stop immediately and let celery return it back to ready queue. Also, from time to time I can notice that task continues to run and even finishes, so celery stores that result in database, however, it does not remove it from `unacked` queue which can cause it to run twice (or more). Anyone noticed similar behavior and possibly can provide some guidance how to fix this?

**Comment by karthikkj172** [comment]
I have the same problem when using celery==5.2.7 with prefork workers.

task_acks_late = True
acks_on_failure_or_timeout = True
worker_lost_wait = 30.
worker_send_task_events = True

[INFO/MainProcess] Task subtask_execution_lib.tasks.cpu_task.cpu_task["my_task_id"] received
[uwsgi] | - SIGINT/SIGTERM received...killing workers...

[uwsgi] |  - [uwsgi-daemons] stopping daemon (pid: 36411): /bin/celery -A worker_a worker --pool=prefork -Ofair --loglevel=info -n celery@local

worker: Cold shutdown (MainProcess)

In my case it remains in started state until I manually revoke
+-----+------------+---------+
| id      | task_id        | status     | 
+-----+------------+---------+
| 230   | my_task_id | STARTED |
+-----+------------+---------+

**Comment by pbudzon** [comment]
See this: https://github.com/celery/kombu/pull/2049 - is your problem possibly related to that race condition?
